README.md
Name
====

HTTP.Request.Provider - Cross-platform HTTP request implementation


SYNOPSIS
========

        // declaring our class
        Class('HTTP.Request.Provider', {
        
        })
        
        // then instantiating it
        var instance = new HTTP.Request.Provider({
        })
        
        // and doing something totally awesome with it :D
        instance.method(param1, param2)
        


DESCRIPTION
===========

`HTTP.Request.Provider` is a stub for Joose-orientied JSAN modules.


ISA
===

None.


DOES
====

None.


TRAITS
======

None.



ATTRIBUTES
==========

### attributeName

> `AttributeType attributeName`

> Attribute description


METHODS
=======

### methodName

> `method signature`

> Method description


EXAMPLES
========

Our class can be used like this:

        // then instantiating it
        var instance = new HTTP.Request.Provider({
        })

and like that:

        // then instantiating it
        var instance = new HTTP.Request.Provider({
        })


GETTING HELP
============

This extension is supported via github issues tracker: <http://github.com/SamuraiJack/Module-Stub/issues>

For general Joose questions you can also visit [#joose](http://webchat.freenode.net/?randomnick=1&channels=joose&prompt=1) 
on irc.freenode.org or the forum at: <http://joose.it/forum>
 


SEE ALSO
========

Web page of this module: <http://github.com/SamuraiJack/Module-Stub/>

General documentation for Joose: <http://openjsan.org/go/?l=Joose>


BUGS
====

All complex software has bugs lurking in it, and this module is no exception.

Please report any bugs through the web interface at <http://github.com/SamuraiJack/Module-Stub/issues>



AUTHORS
=======

Nickolay Platonov <nplatonov@cpan.org>





COPYRIGHT AND LICENSE
=====================

This software is Copyright (c) 2010 by Nickolay Platonov <nplatonov@cpan.org>.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 3, June 2007


Changes
Release history for HTTP.Request.Provider

{{$NEXT}}

0.02      2010-11-01 16:16

        - release for npm

0.01      2010-10-11 16:25

        First version, released on an unsuspecting world.



t/index.js
var Harness
var isNode      = typeof process != 'undefined' && process.pid

if (isNode) {
    require('Task/Test/Run/NodeJSBundle')
    
    Harness = Test.Run.Harness.NodeJS
} else 
    Harness = Test.Run.Harness.Browser.ExtJS
        
    
var INC = (isNode ? require.paths : []).concat('../lib', '/jsan')


Harness.configure({
	title 	: 'HTTP.Request.Provider Test Suite',
    
    verbosity : 1,
    
	preload : [
        "jsan:Task.Joose.Core",
        "jsan:Task.JooseX.Namespace.Depended.Auto",
        {
            text : "JooseX.Namespace.Depended.Manager.my.INC = " + Harness.prepareINC(INC)
        },
        "Task.HTTP.Request.Provider.Auto"
    ]
})


Harness.start(
	'010_sanity.t.js',
    '020_xhr.t.js',
    '030_nodejs.t.js'
)


t/020_xhr.t.js
StartTest(function(t) {
    
    if (Joose.is_NodeJS) {
        t.skip("Tests for browser platform")
        
        t.done()
        
        return
    }
    
    var async0 = t.beginAsync()
    
    use([ 'HTTP.Request.Provider.XHR' ], function () {
        
        //======================================================================================================================================================================================================================================================
        t.diag('GET')
        
        var req = new HTTP.Request.Provider.XHR()
        
        var url = t.harness.absolutizeURL('/jsan/Scope/Provider/static/stub.html')
        
        req.request(url).then(function (res) {
            
            t.ok(res.status == 200, 'Request was successfull')
            t.like(res.text, '<html>', 'And seems it has correct response text')

            t.endAsync(async0)
            
            t.done()
            
        }).now()
        
        
    })
})    

t/010_sanity.t.js
StartTest(function(t) {
    
    var async0 = t.beginAsync()
    
    use([ 'HTTP.Request.Provider', 'HTTP.Request.Provider.XHR', 'HTTP.Request.Provider.NodeJS' ], function () {
        
        //======================================================================================================================================================================================================================================================
        t.diag('Sanity')
        
        t.ok(HTTP.Request.Provider, "HTTP.Request.Provider is here")
        t.ok(HTTP.Request.Provider.XHR, "HTTP.Request.Provider.XHR is here")
        t.ok(HTTP.Request.Provider.NodeJS, "HTTP.Request.Provider.NodeJS is here")
        
        t.endAsync(async0)
        
        t.done()
    })
})    

lib/HTTP/Request/Provider.js
Class('HTTP.Request.Provider', {
    
    trait        : 'JooseX.CPS',
    
    
    has : {
        method          : 'GET',
        
        user            : null,
        password        : null,
        
        url             : null,
        
        query           : null,
        data            : null,
        
        headers         : { is : 'rw' }
    },
    
    
    methods : {
        
        initialize : function () {
            var url     = this.url
            
            if (url) return this.request(url)
        }
    },
    
    
    continued : {
        
        methods : {
            
            request : function (url) {
                throw "Abstract method `request` of the 'HTTP.Request.Provider' reached"
            }
        }
    },
    
    
    my : {
        
        methods : {
        
            getRequest : function (config) {
                
                var providerClass = Joose.is_NodeJS ? HTTP.Request.Provider.NodeJS : HTTP.Request.Provider.XHR
                
                return new providerClass(config)
            }
        }
    }
})

lib/HTTP/Request/Provider/XHR.js
Class('HTTP.Request.Provider.XHR', {
    
    isa        : 'HTTP.Request.Provider',
    
    
    has     : {
        transport : {
            is          : 'rw',
            builder     : 'this.buildTransport'
        }
    },
    
    
    methods : {
        
        buildTransport  : function () {
            if (window.XMLHttpRequest)
                return new XMLHttpRequest()
            else
                return new ActiveXObject("Microsoft.XMLHTTP")
        },
        
        
        urlEncode : function (object) {
            if (typeof object == 'string') return object
            
            if (typeof object == 'object') {
                var queryStr = ''
                
                Joose.O.each(object, function (value, name) {
                    if (queryStr) queryStr += '&'
                    
                    queryStr += encodeURIComponent(name) + '=' + encodeURIComponent(value + '')
                })
                
                return queryStr
            }
            
            return ''
        }
    },
    
    
    continued : {
        
        methods : {
            
            request : function (url) {
                var me          = this
                var CONT        = this.CONT
                var method      = this.method
                
                
                var query       = this.query
                
                if (query) {
                    var encodedQuery = this.urlEncode(query)
                    
                    if (url.match(/\?/))
                        url += '&' + encodedQuery
                    else
                        url += '?' + encodedQuery 
                }
                
                
                var xhr         = this.getTransport()
                
                xhr.open(method, url, true, this.user, this.password)
                
                
                
                var headers     = this.headers
                
                if (headers) Joose.O.eachOwn(headers, function (value, name) {
                    xhr.setRequestHeader(name, value)
                })
                
                
                var data = this.data
                
                if (typeof data == 'object') {
//                    xhr.setRequestHeader("Content-type", "application/x-www-form-urlencoded")
                    
                    data = this.urlEncode(data) 
                }
                
                                
                xhr.onreadystatechange = function () {
                    var status = xhr.status
                    
                    if (xhr.readyState == 4)
                        if (status >= 200 && status < 300) 
                            CONT.CONTINUE({
                                status          : status,
                                text            : xhr.responseText,
                                request         : me
                            })
                        else 
                            CONT.THROW({
                                status          : status,
                                text            : xhr.responseText,
                                request         : me
                            })
                }
                
                xhr.send(data || null)
            }
        }
    }

})


lib/HTTP/Request/Provider/NodeJS.js
Class('HTTP.Request.Provider.NodeJS', {
    
    isa        : 'HTTP.Request.Provider',
    
    
    continued : {
        
        methods : {
            
            request : function (url) {
                var QueryString = require('querystring')
                var CONT        = this.CONT
                
                var parsedURL       = require('url').parse(url)
                parsedURL.pathname  = parsedURL.pathname || '/'
                
                // preparing query part of the URL
                var query       = this.query
                var queryStr    = parsedURL.query || ''
                
                if (typeof query == 'string') queryStr += query
                if (typeof query == 'object') queryStr += (queryStr ? '&' : '') + QueryString.stringify(query)
                
                var requestPath = parsedURL.pathname + (queryStr ? '?' + queryStr : '')
                
                
                var method      = this.method
                var data        = this.data
                var headers     = this.headers || {}
                
                if (data && typeof data == 'object') data = QueryString.stringify(data)
                
                if (!headers.host) {
                    headers.host = parsedURL.hostname
                    
                    if (parsedURL.port) headers.host += ':' + parsedURL.port
                }
                
                if (data) headers[ 'Content-Length' ] = data.length
                
                
                // making the actual request 
                
                var client      = require('http').createClient(parsedURL.port || 80, parsedURL.hostname)
                var request     = client.request(method, requestPath, headers)

                if (data) request.write(data, 'utf8')
                

                request.addListener('response', function (response) {
                    var status          = response.statusCode
                    var buffer          = ''
                    
                    response.setEncoding('utf8')
                    
                    response.addListener('data', function (chunk) {
                        buffer += chunk || ''
                    })
                    
                    response.addListener('end', function () {
                        
                        if (status >= 200 && status < 300)
                            CONT.CONTINUE({
                                status          : status,
                                text            : buffer
                            })
                        else 
                            CONT.THROW({
                                status          : status,
                                text            : buffer
                            })
                    })
                })
                
                request.end()
            }
        }
    }

})
